Index: linux-3.3.8/arch/mips/ath79/machtypes.h
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/machtypes.h	2014-02-17 19:05:26.000000000 +0800
+++ linux-3.3.8/arch/mips/ath79/machtypes.h	2014-02-17 19:08:41.441437714 +0800
@@ -124,6 +124,7 @@
 	ATH79_MACH_ZCN_1523H_5,		/* Zcomax ZCN-1523H-5-xx */
 	ATH79_MACH_NBG6716,		/* ZyXEL NBG6716 */
 	ATH79_MACH_NBG6616,		/* ZyXEL NBG6616 */
+	ATH79_MACH_EMG2926,		/* ZyXEL EMG2926 */	
 };
 
 #endif /* _ATH79_MACHTYPE_H */
Index: linux-3.3.8/arch/mips/ath79/Kconfig
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/Kconfig	2014-02-17 19:05:26.000000000 +0800
+++ linux-3.3.8/arch/mips/ath79/Kconfig	2014-02-17 19:08:00.941236859 +0800
@@ -693,6 +693,23 @@
 	help
 	  Say 'Y' here if you want your kernel to support the
 	  ZyXEL NBG6716.
+	  
+config ATH79_MACH_EMG2926
+	bool "ZyXEL EMG2926"
+	select SOC_QCA955X
+	select ATH79_DEV_ETH
+	select ATH79_DEV_AP9X_PCI if PCI
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_SPI
+	select ATH79_DEV_USB
+	select ATH79_DEV_WMAC
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_NAND
+	select USE_OF
+	help
+	  Say 'Y' here if you want your kernel to support the
+	  ZyXEL EMG2926.	  
 
 endmenu
 
Index: linux-3.3.8/arch/mips/ath79/Makefile
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/Makefile	2014-02-17 19:05:26.000000000 +0800
+++ linux-3.3.8/arch/mips/ath79/Makefile	2014-02-17 19:08:00.941236859 +0800
@@ -103,3 +103,4 @@
 obj-$(CONFIG_ATH79_MACH_WPE72)		+= mach-wpe72.o
 obj-$(CONFIG_ATH79_MACH_WRT160NL)	+= mach-wrt160nl.o
 obj-$(CONFIG_ATH79_MACH_NBG6716)	+= mach-nbg6716.o
+obj-$(CONFIG_ATH79_MACH_EMG2926)	+= mach-emg2926.o
